File: src/scripts/migrate-to-new-structure.ts
/**
 * Frontend Data Migration Utility
 * 
 * This script provides utilities for migrating Firestore data from
 * the old structure to the new enhanced media architecture.
 * 
 * Run from Next.js app:
 *   npm run migrate
 * 
 * Or directly with ts-node:
 *   ts-node src/scripts/migrate-to-new-structure.ts [--dry-run] [--verbose]
 */

import { db } from '../../firebaseConfig';
import { 
  collection, 
  doc, 
  getDocs, 
  getDoc,
  writeBatch,
  updateDoc,
  query,
  where,
  Timestamp
} from 'firebase/firestore';
import { v4 as uuidv4 } from 'uuid';

interface MigrationStats {
  moviesUpdated: number;
  seriesUpdated: number;
  seasonsCreated: number;
  episodesUpdated: number;
  mediaFilesCreated: number;
  assignmentsCreated: number;
  errors: string[];
}

class FrontendDataMigration {
  private db: typeof db;
  private dryRun: boolean;
  private verbose: boolean;
  private stats: MigrationStats;

  constructor(dryRun: boolean = false, verbose: boolean = false) {
    this.dryRun = dryRun;
    this.verbose = verbose;
    this.stats = {
      moviesUpdated: 0,
      seriesUpdated: 0,
      seasonsCreated: 0,
      episodesUpdated: 0,
      mediaFilesCreated: 0,
      assignmentsCreated: 0,
      errors: []
    };

    // Use the existing client-side Firebase configuration
    this.db = db;
  }

  private log(message: string, level: 'info' | 'warn' | 'error' = 'info') {
    if (this.verbose || level === 'error') {
      const prefix = level === 'error' ? '❌' : level === 'warn' ? '⚠️' : '→';
      console.log(`${prefix} ${message}`);
    }
  }

  async migrateAll(): Promise<void> {
    console.log('\n' + '='.repeat(70));
    console.log('MEDIA ARCHITECTURE MIGRATION v2.0 (Frontend)');
    console.log('='.repeat(70));

    if (this.dryRun) {
      console.log('\n⚠️  DRY RUN MODE - No changes will be made');
    }

    console.log('\nStarting migration...\n');

    try {
      // Step 1: Migrate movies
      console.log('📽️  Step 1/5: Migrating movies...');
      await this.migrateMovies();
      console.log(`   ✓ Updated ${this.stats.moviesUpdated} movies`);

      // Step 2: Migrate series and extract seasons
      console.log('\n📺 Step 2/5: Migrating series and extracting seasons...');
      await this.migrateSeries();
      console.log(`   ✓ Updated ${this.stats.seriesUpdated} series`);
      console.log(`   ✓ Created ${this.stats.seasonsCreated} season documents`);

      // Step 3: Migrate episodes
      console.log('\n🎬 Step 3/5: Migrating episodes...');
      await this.migrateEpisodes();
      console.log(`   ✓ Updated ${this.stats.episodesUpdated} episodes`);

      // Step 4: Create media_files
      console.log('\n📁 Step 4/5: Creating media_files collection...');
      await this.createMediaFiles();
      console.log(`   ✓ Created ${this.stats.mediaFilesCreated} media file documents`);

      // Step 5: Create media_assignments
      console.log('\n🔗 Step 5/5: Creating media_assignments...');
      await this.createMediaAssignments();
      console.log(`   ✓ Created ${this.stats.assignmentsCreated} assignment documents`);

      // Summary
      console.log('\n' + '='.repeat(70));
      console.log('MIGRATION COMPLETE');
      console.log('='.repeat(70));
      this.printSummary();

      if (this.dryRun) {
        console.log('\n⚠️  This was a DRY RUN - no actual changes were made');
        console.log('Run without --dry-run to apply changes');
      }

    } catch (error) {
      console.error('\n❌ Migration failed:', error);
      this.stats.errors.push(String(error));
      throw error;
    }
  }

  private async migrateMovies(): Promise<void> {
    try {
      const moviesSnapshot = await getDocs(collection(this.db, 'movies'));
      
      for (const movieDoc of moviesSnapshot.docs) {
        try {
          const movie = movieDoc.data();
          const movieId = movieDoc.id;
          
          this.log(`Migrating movie: ${movie.title || 'Unknown'}`);
          
          const updates: any = {};
          
          // Initialize assignmentSummary if has libraryFiles
          const libraryFiles = movie.libraryFiles || [];
          if (libraryFiles.length > 0) {
            const totalSize = libraryFiles.reduce((sum: number, f: any) => sum + (f.size || 0), 0);
            updates.assignmentSummary = {
              totalFiles: libraryFiles.length,
              versions: [],
              hasPhysicalCopy: movie.physicalCopy || false,
              totalFileSize: totalSize,
              lastUpdated: new Date().toISOString()
            };
          }
          
          // Initialize JellyfinInfo
          updates.jellyfinInfo = {
            folderId: null,
            folderPath: null,
            folderName: null,
            isOrganized: false
          };
          
          // Initialize ExternalIds
          if (!movie.externalIds) {
            updates.externalIds = {
              imdbId: movie.imdbId || null,
              tmdbId: movie.tmdbId || null,
              rottenTomatoesId: null,
              metacriticId: null,
              letterboxdId: null
            };
          }
          
          // Add content rating
          if (!movie.contentRating && movie.rated) {
            updates.contentRating = {
              country: 'US',
              rating: movie.rated,
              ratingSystem: 'MPAA'
            };
          }
          
          if (Object.keys(updates).length > 0 && !this.dryRun) {
            const movieRef = doc(this.db, 'movies', movieId);
            const batch = writeBatch(this.db);
            batch.update(movieRef, updates);
            await batch.commit();
          }
          
          this.stats.moviesUpdated++;
          
        } catch (error) {
          const errorMsg = `Failed to migrate movie ${movieDoc.id}: ${error}`;
          this.log(errorMsg, 'error');
          this.stats.errors.push(errorMsg);
        }
      }
    } catch (error) {
      console.error('Movie migration failed:', error);
      throw error;
    }
  }

  private async migrateSeries(): Promise<void> {
    try {
      const seriesSnapshot = await getDocs(collection(this.db, 'series'));
      
      for (const seriesDoc of seriesSnapshot.docs) {
        try {
          const series = seriesDoc.data();
          const seriesId = seriesDoc.id;
          
          this.log(`Migrating series: ${series.title || 'Unknown'}`);
          
          const seasons = series.seasons || [];
          const seasonIds: string[] = [];
          
          // Create standalone season documents
          for (const seasonData of seasons) {
            const seasonId = uuidv4();
            seasonIds.push(seasonId);
            
            const seasonDoc = {
              id: seasonId,
              seriesId: seriesId,
              seriesTitle: series.title,
              seasonNumber: seasonData.seasonNumber || 0,
              title: seasonData.title || `Season ${seasonData.seasonNumber || 0}`,
              overview: seasonData.overview || null,
              airDate: seasonData.airDate || null,
              episodeCount: seasonData.episodeCount || 0,
              episodeIds: [],
              posterPath: seasonData.posterPath || null,
              episodesWithFiles: 0,
              totalFiles: 0,
              totalFileSize: 0,
              jellyfinFolderId: null,
              jellyfinFolderName: null,
              dateCreated: new Date().toISOString(),
              dateUpdated: new Date().toISOString()
            };
            
            if (!this.dryRun) {
              const batch = writeBatch(this.db);
              const seasonRef = doc(this.db, 'seasons', seasonId);
              batch.set(seasonRef, seasonDoc);
              await batch.commit();
            }
            
            this.stats.seasonsCreated++;
          }
          
          // Update series document
          const updates: any = {
            seriesSummary: {
              totalSeasons: seasons.length,
              totalEpisodes: seasons.reduce((sum: number, s: any) => sum + (s.episodeCount || 0), 0),
              totalRuntime: 0,
              status: series.status || 'Unknown',
              lastAirDate: series.lastAirDate || null
            },
            seriesAssignmentSummary: {
              seasonsWithFiles: 0,
              episodesWithFiles: 0,
              totalFiles: 0,
              totalFileSize: 0,
              lastUpdated: new Date().toISOString()
            },
            seriesJellyfinInfo: {
              folderId: null,
              folderPath: null,
              seasonFolders: [],
              isOrganized: false
            },
            seasonIds: seasonIds
          };
          
          if (!series.externalIds) {
            updates.externalIds = {
              imdbId: series.imdbId || null,
              tmdbId: series.tmdbId || null,
              tvdbId: series.tvdbId || null,
              tvMazeId: null
            };
          }
          
          if (!this.dryRun) {
            const batch = writeBatch(this.db);
            const seriesRef = doc(this.db, 'series', seriesId);
            batch.update(seriesRef, updates);
            await batch.commit();
          }
          
          this.stats.seriesUpdated++;
          
        } catch (error) {
          const errorMsg = `Failed to migrate series ${seriesDoc.id}: ${error}`;
          this.log(errorMsg, 'error');
          this.stats.errors.push(errorMsg);
        }
      }
    } catch (error) {
      console.error('Series migration failed:', error);
      throw error;
    }
  }

  private async migrateEpisodes(): Promise<void> {
    try {
      const episodesSnapshot = await getDocs(collection(this.db, 'episodes'));
      
      for (const episodeDoc of episodesSnapshot.docs) {
        try {
          const episode = episodeDoc.data();
          
          const updates: any = {
            hasFile: false,
            fileId: null,
            fileCount: 0,
            jellyfinFilename: null
          };
          
          if (!episode.externalIds) {
            updates.externalIds = {
              imdbId: episode.imdbId || null,
              tmdbId: episode.tmdbId || null,
              tvdbId: episode.tvdbId || null
            };
          }
          
          if (Object.keys(updates).length > 0 && !this.dryRun) {
            const episodeRef = doc(this.db, 'episodes', episodeDoc.id);
            await updateDoc(episodeRef, updates);
          }
          
          this.stats.episodesUpdated++;
          
        } catch (error) {
          const errorMsg = `Failed to migrate episode ${episodeDoc.id}: ${error}`;
          this.log(errorMsg, 'error');
          this.stats.errors.push(errorMsg);
        }
      }
    } catch (error) {
      console.error('Episode migration failed:', error);
      throw error;
    }
  }

  private async createMediaFiles(): Promise<void> {
    try {
      const moviesSnapshot = await getDocs(collection(this.db, 'movies'));
      
      for (const movieDoc of moviesSnapshot.docs) {
        const movie = movieDoc.data();
        const libraryFiles = movie.libraryFiles || [];
        
        if (libraryFiles.length === 0) continue;
        
        for (const fileData of libraryFiles) {
          try {
            const fileId = uuidv4();
            
            const fileDoc = {
              id: fileId,
              userId: movie.userId || 'migrated',
              libraryPath: movie.libraryPath || '',
              filePath: fileData.path || '',
              fileName: fileData.name || '',
              fileExtension: fileData.extension || '',
              fileSize: fileData.size || 0,
              dateAdded: new Date().toISOString(),
              dateModified: fileData.modified || new Date().toISOString(),
              dateCreated: fileData.created || new Date().toISOString(),
              containerFormat: null,
              duration: null,
              overallBitrate: null,
              videoMetadata: null,
              audioTracks: [],
              subtitleTracks: [],
              checksum: null,
              isAssigned: true,
              assignmentId: null,
              mediaType: 'movie',
              mediaId: movieDoc.id,
              parsedTitle: null,
              parsedYear: null,
              parsedQuality: null,
              releaseGroup: null
            };
            
            if (!this.dryRun) {
              const batch = writeBatch(this.db);
              const fileRef = doc(this.db, 'media_files', fileId);
              batch.set(fileRef, fileDoc);
              await batch.commit();
            }
            
            this.stats.mediaFilesCreated++;
            
          } catch (error) {
            const errorMsg = `Failed to create media_file: ${error}`;
            this.log(errorMsg, 'error');
            this.stats.errors.push(errorMsg);
          }
        }
      }
    } catch (error) {
      console.error('Media files creation failed:', error);
      throw error;
    }
  }

  private async createMediaAssignments(): Promise<void> {
    try {
      const filesQuery = query(
        collection(this.db, 'media_files'),
        where('isAssigned', '==', true),
        where('assignmentId', '==', null)
      );
      const filesSnapshot = await getDocs(filesQuery);
      
      for (const fileDoc of filesSnapshot.docs) {
        try {
          const file = fileDoc.data();
          const assignmentId = uuidv4();
          
          const assignmentDoc = {
            id: assignmentId,
            userId: file.userId || 'migrated',
            fileId: fileDoc.id,
            mediaType: file.mediaType || 'movie',
            mediaId: file.mediaId,
            version: '1080p',
            status: 'assigned',
            dateAssigned: new Date().toISOString(),
            isOrganized: false,
            targetFolder: null,
            sourceFile: {
              filePath: file.filePath,
              fileName: file.fileName,
              fileSize: file.fileSize || 0
            },
            organizationHistory: []
          };
          
          if (!this.dryRun) {
            const batch = writeBatch(this.db);
            const assignmentRef = doc(this.db, 'media_assignments', assignmentId);
            batch.set(assignmentRef, assignmentDoc);
            
            const fileRef = doc(this.db, 'media_files', fileDoc.id);
            batch.update(fileRef, {
              assignmentId,
              organizationStatus: 'pending',
              updatedAt: new Date().toISOString()
            });
            
            await batch.commit();
          }
          
          this.stats.assignmentsCreated++;
          
        } catch (error) {
          const errorMsg = `Failed to create assignment: ${error}`;
          this.log(errorMsg, 'error');
          this.stats.errors.push(errorMsg);
        }
      }
    } catch (error) {
      console.error('Assignment creation failed:', error);
      throw error;
    }
  }

  private printSummary(): void {
    console.log('\nMigration Statistics:');
    console.log(`  Movies updated:          ${this.stats.moviesUpdated}`);
    console.log(`  Series updated:          ${this.stats.seriesUpdated}`);
    console.log(`  Seasons created:         ${this.stats.seasonsCreated}`);
    console.log(`  Episodes updated:        ${this.stats.episodesUpdated}`);
    console.log(`  Media files created:     ${this.stats.mediaFilesCreated}`);
    console.log(`  Assignments created:     ${this.stats.assignmentsCreated}`);
    
    if (this.stats.errors.length > 0) {
      console.log(`\n⚠️  Errors encountered:     ${this.stats.errors.length}`);
      if (this.verbose) {
        console.log('\nError details:');
        this.stats.errors.slice(0, 10).forEach(error => {
          console.log(`  - ${error}`);
        });
        if (this.stats.errors.length > 10) {
          console.log(`  ... and ${this.stats.errors.length - 10} more`);
        }
      }
    }
  }
}

// Main execution
async function main() {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run');
  const verbose = args.includes('--verbose') || args.includes('-v');

  const migration = new FrontendDataMigration(dryRun, verbose);
  
  try {
    await migration.migrateAll();
    console.log('\n✅ Migration completed successfully!');
    process.exit(0);
  } catch (error) {
    console.error('\n❌ Migration failed:', error);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

export { FrontendDataMigration };

File: backend/scripts/migrate_to_new_structure.py
"""
Data Migration Script for Media Architecture v2

This script migrates existing data from the old structure to the new
enhanced media architecture with Jellyfin integration.

Migrations:
1. Movies: Add new fields, deprecate old fields, preserve data
2. Series: Extract seasons to standalone collection
3. Create media_files from existing libraryFiles
4. Create media_assignments for existing file links
5. Initialize assignment summaries

Usage:
    python migrate_to_new_structure.py [--dry-run] [--verbose]

Author: Media Manager Team
Version: 2.0.0
"""

import argparse
import sys
from datetime import datetime
from typing import Dict, List, Any, Optional
import uuid

# Add parent directory to path for imports
sys.path.append('..')

from services.firestore_service import FirestoreService
from config.settings import settings
from utils.logging import logger


class DataMigration:
    """Handles migration of existing data to new architecture"""
    
    def __init__(self, dry_run: bool = False, verbose: bool = False):
        """
        Initialize migration service.
        
        Args:
            dry_run: If True, simulate migration without making changes
            verbose: If True, log detailed progress
        """
        self.dry_run = dry_run
        self.verbose = verbose
        self.firestore = FirestoreService(settings.firebase_project_id)
        self.stats = {
            'movies_updated': 0,
            'series_updated': 0,
            'seasons_created': 0,
            'episodes_updated': 0,
            'media_files_created': 0,
            'assignments_created': 0,
            'errors': []
        }
    
    def log(self, message: str, level: str = 'info'):
        """Log message if verbose mode enabled"""
        if self.verbose:
            if level == 'info':
                logger.info(message)
            elif level == 'warning':
                logger.warning(message)
            elif level == 'error':
                logger.error(message)
    
    def migrate_all(self):
        """Run all migration steps in correct order"""
        print("\n" + "="*70)
        print("MEDIA ARCHITECTURE MIGRATION v2.0")
        print("="*70)
        
        if self.dry_run:
            print("\n⚠️  DRY RUN MODE - No changes will be made")
        
        print("\nStarting migration...\n")
        
        try:
            # Step 1: Migrate movies
            print("📽️  Step 1/5: Migrating movies...")
            self.migrate_movies()
            print(f"   ✓ Updated {self.stats['movies_updated']} movies")
            
            # Step 2: Migrate series and extract seasons
            print("\n📺 Step 2/5: Migrating series and extracting seasons...")
            self.migrate_series()
            print(f"   ✓ Updated {self.stats['series_updated']} series")
            print(f"   ✓ Created {self.stats['seasons_created']} season documents")
            
            # Step 3: Migrate episodes
            print("\n🎬 Step 3/5: Migrating episodes...")
            self.migrate_episodes()
            print(f"   ✓ Updated {self.stats['episodes_updated']} episodes")
            
            # Step 4: Create media_files from libraryFiles
            print("\n📁 Step 4/5: Creating media_files collection...")
            self.create_media_files()
            print(f"   ✓ Created {self.stats['media_files_created']} media file documents")
            
            # Step 5: Create media_assignments
            print("\n🔗 Step 5/5: Creating media_assignments...")
            self.create_media_assignments()
            print(f"   ✓ Created {self.stats['assignments_created']} assignment documents")
            
            # Summary
            print("\n" + "="*70)
            print("MIGRATION COMPLETE")
            print("="*70)
            self.print_summary()
            
            if self.dry_run:
                print("\n⚠️  This was a DRY RUN - no actual changes were made")
                print("Run without --dry-run to apply changes")
            
        except Exception as e:
            logger.error("Migration failed", error=str(e))
            print(f"\n❌ Migration failed: {str(e)}")
            self.stats['errors'].append(str(e))
            raise
    
    def migrate_movies(self):
        """Migrate movie documents to new structure"""
        try:
            # Get all movies
            movies = self.firestore.get_collection('movies')
            
            for movie in movies:
                try:
                    movie_id = movie.get('id')
                    if not movie_id:
                        continue
                    
                    self.log(f"Migrating movie: {movie.get('title', 'Unknown')}", 'info')
                    
                    # Prepare update data
                    updates = {}
                    
                    # Initialize assignmentSummary if has libraryFiles
                    library_files = movie.get('libraryFiles', [])
                    if library_files:
                        total_size = sum(f.get('size', 0) for f in library_files if isinstance(f, dict))
                        updates['assignmentSummary'] = {
                            'totalFiles': len(library_files),
                            'versions': [],  # Will be populated during assignment
                            'hasPhysicalCopy': movie.get('physicalCopy', False),
                            'totalFileSize': total_size,
                            'lastUpdated': datetime.utcnow().isoformat()
                        }
                    
                    # Initialize JellyfinInfo
                    updates['jellyfinInfo'] = {
                        'folderId': None,
                        'folderPath': None,
                        'folderName': None,
                        'isOrganized': False
                    }
                    
                    # Initialize ExternalIds if not present
                    if 'externalIds' not in movie:
                        updates['externalIds'] = {
                            'imdbId': movie.get('imdbId'),
                            'tmdbId': movie.get('tmdbId'),
                            'rottenTomatoesId': None,
                            'metacriticId': None,
                            'letterboxdId': None
                        }
                    
                    # Add content rating structure
                    if 'contentRating' not in movie and movie.get('rated'):
                        updates['contentRating'] = {
                            'country': 'US',
                            'rating': movie.get('rated'),
                            'ratingSystem': 'MPAA'
                        }
                    
                    # Apply updates
                    if updates and not self.dry_run:
                        self.firestore.update_document('movies', movie_id, updates)
                    
                    self.stats['movies_updated'] += 1
                    
                except Exception as e:
                    error_msg = f"Failed to migrate movie {movie.get('id')}: {str(e)}"
                    self.log(error_msg, 'error')
                    self.stats['errors'].append(error_msg)
                    continue
                    
        except Exception as e:
            logger.error("Movie migration failed", error=str(e))
            raise
    
    def migrate_series(self):
        """Migrate series and extract seasons to standalone collection"""
        try:
            # Get all series
            series_list = self.firestore.get_collection('series')
            
            for series in series_list:
                try:
                    series_id = series.get('id')
                    if not series_id:
                        continue
                    
                    self.log(f"Migrating series: {series.get('title', 'Unknown')}", 'info')
                    
                    # Extract seasons array
                    seasons = series.get('seasons', [])
                    season_ids = []
                    
                    # Create standalone season documents
                    for season_data in seasons:
                        if not isinstance(season_data, dict):
                            continue
                        
                        season_id = str(uuid.uuid4())
                        season_ids.append(season_id)
                        
                        # Create season document
                        season_doc = {
                            'id': season_id,
                            'seriesId': series_id,
                            'seriesTitle': series.get('title'),
                            'seasonNumber': season_data.get('seasonNumber', 0),
                            'title': season_data.get('title', f"Season {season_data.get('seasonNumber', 0)}"),
                            'overview': season_data.get('overview'),
                            'airDate': season_data.get('airDate'),
                            'episodeCount': season_data.get('episodeCount', 0),
                            'episodeIds': [],  # Will be populated when migrating episodes
                            'posterPath': season_data.get('posterPath'),
                            'episodesWithFiles': 0,
                            'totalFiles': 0,
                            'totalFileSize': 0,
                            'jellyfinFolderId': None,
                            'jellyfinFolderName': None,
                            'dateCreated': datetime.utcnow().isoformat(),
                            'dateUpdated': datetime.utcnow().isoformat()
                        }
                        
                        if not self.dry_run:
                            self.firestore.create_document('seasons', season_id, season_doc)
                        
                        self.stats['seasons_created'] += 1
                    
                    # Update series document
                    updates = {
                        'seriesSummary': {
                            'totalSeasons': len(seasons),
                            'totalEpisodes': sum(s.get('episodeCount', 0) for s in seasons if isinstance(s, dict)),
                            'totalRuntime': 0,  # Will be calculated from episodes
                            'status': series.get('status', 'Unknown'),
                            'lastAirDate': series.get('lastAirDate')
                        },
                        'seriesAssignmentSummary': {
                            'seasonsWithFiles': 0,
                            'episodesWithFiles': 0,
                            'totalFiles': 0,
                            'totalFileSize': 0,
                            'lastUpdated': datetime.utcnow().isoformat()
                        },
                        'seriesJellyfinInfo': {
                            'folderId': None,
                            'folderPath': None,
                            'seasonFolders': [],
                            'isOrganized': False
                        },
                        'seasonIds': season_ids
                    }
                    
                    # Initialize external IDs
                    if 'externalIds' not in series:
                        updates['externalIds'] = {
                            'imdbId': series.get('imdbId'),
                            'tmdbId': series.get('tmdbId'),
                            'tvdbId': series.get('tvdbId'),
                            'tvMazeId': None
                        }
                    
                    if not self.dry_run:
                        self.firestore.update_document('series', series_id, updates)
                    
                    self.stats['series_updated'] += 1
                    
                except Exception as e:
                    error_msg = f"Failed to migrate series {series.get('id')}: {str(e)}"
                    self.log(error_msg, 'error')
                    self.stats['errors'].append(error_msg)
                    continue
                    
        except Exception as e:
            logger.error("Series migration failed", error=str(e))
            raise
    
    def migrate_episodes(self):
        """Migrate episode documents"""
        try:
            episodes = self.firestore.get_collection('episodes')
            
            for episode in episodes:
                try:
                    episode_id = episode.get('id')
                    if not episode_id:
                        continue
                    
                    updates = {
                        'hasFile': False,
                        'fileId': None,
                        'fileCount': 0,
                        'jellyfinFilename': None
                    }
                    
                    # Initialize external IDs
                    if 'externalIds' not in episode:
                        updates['externalIds'] = {
                            'imdbId': episode.get('imdbId'),
                            'tmdbId': episode.get('tmdbId'),
                            'tvdbId': episode.get('tvdbId')
                        }
                    
                    if not self.dry_run:
                        self.firestore.update_document('episodes', episode_id, updates)
                    
                    self.stats['episodes_updated'] += 1
                    
                except Exception as e:
                    error_msg = f"Failed to migrate episode {episode.get('id')}: {str(e)}"
                    self.log(error_msg, 'error')
                    self.stats['errors'].append(error_msg)
                    continue
                    
        except Exception as e:
            logger.error("Episode migration failed", error=str(e))
            raise
    
    def create_media_files(self):
        """Create media_files collection from existing libraryFiles"""
        try:
            # Get all movies with libraryFiles
            movies = self.firestore.get_collection('movies')
            
            for movie in movies:
                library_files = movie.get('libraryFiles', [])
                if not library_files:
                    continue
                
                for file_data in library_files:
                    if not isinstance(file_data, dict):
                        continue
                    
                    try:
                        file_id = str(uuid.uuid4())
                        
                        file_doc = {
                            'id': file_id,
                            'userId': movie.get('userId', 'migrated'),
                            'libraryPath': movie.get('libraryPath', ''),
                            'filePath': file_data.get('path', ''),
                            'fileName': file_data.get('name', ''),
                            'fileExtension': file_data.get('extension', ''),
                            'fileSize': file_data.get('size', 0),
                            'dateAdded': datetime.utcnow().isoformat(),
                            'dateModified': file_data.get('modified', datetime.utcnow().isoformat()),
                            'dateCreated': file_data.get('created', datetime.utcnow().isoformat()),
                            'containerFormat': None,
                            'duration': None,
                            'overallBitrate': None,
                            'videoMetadata': None,
                            'audioTracks': [],
                            'subtitleTracks': [],
                            'checksum': None,
                            'isAssigned': True,  # Already linked to movie
                            'assignmentId': None,  # Will be set when creating assignments
                            'mediaType': 'movie',
                            'mediaId': movie.get('id'),
                            'parsedTitle': None,
                            'parsedYear': None,
                            'parsedQuality': None,
                            'releaseGroup': None
                        }
                        
                        if not self.dry_run:
                            self.firestore.create_document('media_files', file_id, file_doc)
                        
                        self.stats['media_files_created'] += 1
                        
                    except Exception as e:
                        error_msg = f"Failed to create media_file: {str(e)}"
                        self.log(error_msg, 'error')
                        self.stats['errors'].append(error_msg)
                        continue
                        
        except Exception as e:
            logger.error("Media files creation failed", error=str(e))
            raise
    
    def create_media_assignments(self):
        """Create media_assignments for existing file links"""
        try:
            # Get all media_files that are assigned but have no assignmentId
            media_files = self.firestore.get_collection('media_files')
            
            for file_doc in media_files:
                if not file_doc.get('isAssigned') or file_doc.get('assignmentId'):
                    continue
                
                try:
                    assignment_id = str(uuid.uuid4())
                    
                    assignment_doc = {
                        'id': assignment_id,
                        'userId': file_doc.get('userId', 'migrated'),
                        'fileId': file_doc['id'],
                        'mediaType': file_doc.get('mediaType', 'movie'),
                        'mediaId': file_doc.get('mediaId'),
                        'version': '1080p',  # Default version
                        'status': 'assigned',
                        'dateAssigned': datetime.utcnow().isoformat(),
                        'isOrganized': False,
                        'targetFolder': None,
                        'sourceFile': {
                            'filePath': file_doc.get('filePath'),
                            'fileName': file_doc.get('fileName'),
                            'fileSize': file_doc.get('fileSize', 0)
                        },
                        'organizationHistory': []
                    }
                    
                    if not self.dry_run:
                        self.firestore.create_document('media_assignments', assignment_id, assignment_doc)
                        # Update media_file with assignmentId
                        self.firestore.update_document('media_files', file_doc['id'], {
                            'assignmentId': assignment_id
                        })
                    
                    self.stats['assignments_created'] += 1
                    
                except Exception as e:
                    error_msg = f"Failed to create assignment: {str(e)}"
                    self.log(error_msg, 'error')
                    self.stats['errors'].append(error_msg)
                    continue
                    
        except Exception as e:
            logger.error("Assignment creation failed", error=str(e))
            raise
    
    def print_summary(self):
        """Print migration summary statistics"""
        print("\nMigration Statistics:")
        print(f"  Movies updated:          {self.stats['movies_updated']}")
        print(f"  Series updated:          {self.stats['series_updated']}")
        print(f"  Seasons created:         {self.stats['seasons_created']}")
        print(f"  Episodes updated:        {self.stats['episodes_updated']}")
        print(f"  Media files created:     {self.stats['media_files_created']}")
        print(f"  Assignments created:     {self.stats['assignments_created']}")
        
        if self.stats['errors']:
            print(f"\n⚠️  Errors encountered:     {len(self.stats['errors'])}")
            if self.verbose:
                print("\nError details:")
                for error in self.stats['errors'][:10]:  # Show first 10
                    print(f"  - {error}")
                if len(self.stats['errors']) > 10:
                    print(f"  ... and {len(self.stats['errors']) - 10} more")


def main():
    """Main entry point for migration script"""
    parser = argparse.ArgumentParser(description='Migrate media data to new architecture v2')
    parser.add_argument('--dry-run', action='store_true', help='Simulate migration without making changes')
    parser.add_argument('--verbose', '-v', action='store_true', help='Enable verbose logging')
    
    args = parser.parse_args()
    
    # Create migration instance
    migration = DataMigration(dry_run=args.dry_run, verbose=args.verbose)
    
    try:
        # Run migration
        migration.migrate_all()
        
        print("\n✅ Migration completed successfully!")
        sys.exit(0)
        
    except Exception as e:
        print(f"\n❌ Migration failed: {str(e)}")
        sys.exit(1)


if __name__ == '__main__':
    main()
